cpp/dolfinx/mesh/MeshTags.h

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2020 Michal Habera
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include <dolfinx/common/IndexMap.h>
+#include "Mesh.h"
+#include "Topology.h"
+#include <map>
+#include <memory>
+#include <utility>
+
+namespace dolfinx
+{
+namespace mesh
+{
+
+/// A MeshTags is a class used to tag mesh entities using their
+/// local-to-process index and an attached value.
+/// MeshTags is a sparse data storage class, since it allows to
+/// tag only few mesh entities.
+/// @tparam Type
+template <typename T>
+class MeshTags
+{
+public:
+  /// Create from entities of given dimension on a mesh
+  /// @param[in] mesh The mesh associated with the tags
+  /// @param[in] dim Topological dimension of mesh entities
+  ///    to tag.
+  /// @param[in] indices Array of indices, will be copied.
+  ///    Local-to-process.
+  /// @param[in] values Array of values attached to indices,
+  ///    will be copied.
+  MeshTags(std::shared_ptr<const Mesh> mesh, int dim,
+           const Eigen::Array<std::int32_t, Eigen::Dynamic, 1>& indices,
+           const Eigen::Array<T, Eigen::Dynamic, 1>& values);
+
+  /// Destructor
+  ~MeshTags() = default;
+
+  /// Associated mesh
+  const std::shared_ptr<const Mesh> mesh;
+
+  /// Topological dimension of tagged mesh entities
+  const int dim;
```

Comment:
Make this a method - direct access to `const` member data prevents move semantics.